### ISSUE TITLE
[bug] cannot create package with tool_requires(cmake)

### ISSUE THREAD

**Issue by paulharris** [conan-io/conan#11287]
See conan-center-index#10833

```
    def requirements(self):
        self.requires("libcurl/7.80.0")
        self.requires("openssl/1.1.1o", override=True)

    def build_requirements(self):
        self.tool_requires("cmake/3.22.4")
```

Fails with:
```
ERROR: Conflict in cmake/3.22.4:
    'cmake/3.22.4' requires 'openssl/1.1.1n' while 'libcurl/7.80.0' requires 'openssl/1.1.1o'.
    To fix this conflict you need to override the package 'openssl' in your root package.
```

**Comment by memsharded** [comment]
I think this has been addressed in https://github.com/conan-io/conan-center-index/pull/10833, this is not a Conan issue per se, but the fact that ConanCenter is still not using the 2 host/build profiles yet. Using the two profiles would solve the issue.

**Comment by paulharris** [comment]
This happens locally to me as well, not just on CCI's CI build platform.
Can I somehow use 2 host/build profiles locally?  How?   Is that -pr:b=default -pr:h=default or something like that?

Is the solution that we must bump the openssl version across the board all the time?

Is this problem related to OpenSSL's weird version number, ie 1.1.1n instead 1.1.2... is that semver compliant?

**Comment by SpaceIm** [comment]
To always use 2 profiles by default, you can create `global.conf` file in `.conan` folder with this content:

```ini
core:default_profile=default
core:default_build_profile=default
```

> Is the solution that we must bump the openssl version across the board all the time?

Even with 2 profiles, conflicts may occur in conan-center recipes since there is no version range. Consumers can resolve these conflicts, but not CI of conan-center-index. This is why, sometimes, widely used dependencies (mainly zlib, openssl and boost, but also libcurl, libiconv, freetype, expat, libxml2, xz_utils, sqlite3, libpng, libjpeg and libtiff) are massively bumped in all recipes when a new version is released. It's a tedious process, and CCI policy regarding dependencies versions doesn't scale very well (one new version of zlib or openssl means ~100 recipes to update).

**Comment by paulharris** [comment]
Does CCI need the human in the loop to check each dependency bump will work?
I can understand that you want a human to "sign off" on a new (eg) openssl version before applying changes to every consumer of CCI packages,

But would it work if there was an external list of "compatible versions" that a human added to, and these recipes referred to?
ie the requires = "openssl/1.1.1k or CCI compatible"   ie a special kind of version range,
and some OTHER recipe in CCI (that the consumer could fork+hack if required) would list versions of packages that are backward compatible, and forward compatible

eg
1.1.1k can upgrade to [1.1.1l, 1.1.1m, 1.1.1n, ... ]
1.1.1n can downgrade to [1.1.1m]  --- ie not .l and .k

and for simpler situations, where it doesn't matter
all equivalent both ways [1.1.1k, 1.1.1l, 1.1.1m, 1.1.1n, 1.1.1o]

?
Then updating one recipe would result in many updating automatically?

**Comment by alwaysmpe** [comment]
Is it actually possible to override build requirement sub dependencies? As an example, I'm trying to use python and cmake:

    from conans import ConanFile

    class TestConflict(ConanFile):

        def build_requirements(self):
            self.tool_requires("openssl/1.1.1o")
            self.tool_requires("cpython/3.9.7")
            self.tool_requires("cmake/3.22.4")

This gives me the error:

    ERROR: Conflict in cpython/3.9.7:
        'cpython/3.9.7' requires 'openssl/1.1.1l' while 'conanfile.py' requires 'openssl/1.1.1o'.
        To fix this conflict you need to override the package 'openssl' in your root package.

This is the base package I'm trying to install. Removing the openssl declaration gives a different conflict between cpython and cmake having different openssl versions. I've got conan v2 enabled & 2 profiles. `tool_requires` doesn't accept an override argument.

I'm aware I could create an updated recipe and push to cache/private artifactory repo, but this takes time/seems like overkill for what is likely to be a common problem.

These aren't the actual recipes we're using, but they do demonstrate the problem, we're using a mix of public/private recipes.

**Comment by memsharded** [comment]
In Conan 1.X it is not possible to override transitive dependencies of ``tool_requires`` (or ``build_requires``, because they are kind of invisible to the consumer). Explicitly depending on them with ``self.tool_requires("openssl/1.1.1o")`` doesn't solve it, as it just introduce a direct and conflicting ``tool_requires`` to openssl, which doesn't make much sense.

Conan 2.0 will allow overriding dependencies in the "build" context, as long as they are declared ``visible=True``, otherwise they will keep private. 

In any case, the idea is that the dependencies of ``cmake`` package should be irrelevant (and ``tool_requires`` in general), in the same way that they are if you download the precompiled binary for cmake from the CMake web page. If an updated version of cmake->openssl dependency is wanted, then that should be updated in the ``cmake`` package recipe.

**Comment by alwaysmpe** [comment]
I think understand what you're saying.

I think the problem is that in the case that prebuilts aren't available, then conflicting requirements of different `tool_requires` should, presumably, be ignored, where currently they aren't? And, as such, conflicts result in failures. My take away is "wait for v2 to be fully released before trying too much with this use case"? Happy with that, will wait for updates.

I'm also assuming there's nothing to do to enable v2 features beyond `export CONAN_V2_MODE=1`?

**Comment by memsharded** [comment]
> I think the problem is that in the case that prebuilts aren't available, then conflicting requirements of different tool_requires should, presumably, be ignored, where currently they aren't? And, as such, conflicts result in failures. 

Yes, Conan 1.X might have some scenarios creating conflicts, that can be more easy to produce if needing to build from sources ``tool_requires`` too. Also, in general it is very recommended to use the 2 profiles approach, with ``--profile:build=xxx --profile:host=yyyy``, that will minimize conflicts with tool_requires.

``CONAN_V2_MODE`` was just to do some simple checks, not really changing behavior. Using the 2 profiles is actually changing the behavior to be more 2.0 aligned.

**Comment by memsharded** [comment]
Closing as outdated, solved long time ago.
Please create new tickets for any further question or issue, thanks for the feedback!
